### PR TITLE
Experimental transpilation functionality for nested worklets

### PR DESCRIPTION
**Pull request by tjzel**
## Summary

This PR adds optional functionality to plugin that allows nesting worklets. This is necessary to be able to call functions in-between threads and is a step forward to allow multi-threading in react-native-reanimated.

Current use cases are provided in the comment below.

## Test plan

Added some test suites.

### CODE REVIEW

File: plugin/src/types.ts [hunk 1]
```diff
@@ -6,6 +6,7 @@ export interface ReanimatedPluginPass {
   opts: {
     relativeSourceLocation?: boolean;
     disableInlineStylesWarning?: boolean;
+    useOnExitLogicForWorkletNodes?: boolean;
```
Comment (tomekzaw) [inline plugin/src/types.ts:9]: Good names describe what is done, not how. Please consider renaming this option to something like `processNestedWorklets` or similar.

File: plugin/src/processIfWorkletNode.ts [hunk 2]
```diff
@@ -15,30 +15,41 @@ export function processIfWorkletNode(
...
+          const directives = fun.node.body.directives;
+          if (
+            directives &&
+            directives.length > 0 &&
+            directives.some(
+              (directive) =>
+                isDirectiveLiteral(directive.value) &&
+                directive.value.value === 'worklet'
+            )
+          ) {
```
Comment (tomekzaw) [inline plugin/src/processIfWorkletNode.ts:37]: Can we move this logic to a separate function `hasWorkletDirective(fun)`?

File: plugin/src/processIfWorkletNode.ts [hunk 3]
```diff
@@ -15,30 +15,41 @@ export function processIfWorkletNode(
...
+            shouldBeProcessed = true;
+          }
+        } else if (
+          state.opts.useOnExitLogicForWorkletNodes &&
+          (parent?.isFunctionDeclaration() ||
+            parent?.isFunctionExpression() ||
+            parent?.isArrowFunctionExpression())
```
Comment (tomekzaw) [inline plugin/src/processIfWorkletNode.ts:44]: Can we move this logic to a separate function `isWorkletizableFunction(parent)`?

**Comment by tomekzaw** [comment]
Use case:

```tsx
runOnUI(() => {
  console.log('Hello from UI thread');
  runOnJS(() => {
    'worklet';
    console.log('Hello from JS thread');
    runOnUI(() => {
      console.log('Hello from UI thread again');
    })();
  })();
})();
```

```tsx
runOnUI(() => {
  const a = () => {
    'worklet';
    console.log('Good morning from JS thread!');
  };
  const b = () => {
    'worklet';
    console.log('Good after noon from JS thread');
  };
  const func = Math.random() < 0.5 ? a : b;
  runOnJS(func)();
})();
```

File: __tests__/plugin.test.ts [hunk 4]
```diff
@@ -1442,4 +1412,163 @@ describe('babel plugin', () => {
       expect(code).toMatchSnapshot();
     });
   });
+
+  describe('for worklet nesting', () => {
+    it("doesn't nest worklets for other threads when disabled", () => {
```
Comment (tomekzaw) [inline __tests__/plugin.test.ts:1417]:
```suggestion
    it("doesn't process nested worklets when disabled", () => {
```

File: __tests__/plugin.test.ts [hunk 5]
```diff
@@ -1442,4 +1412,163 @@ describe('babel plugin', () => {
...
+          bar();
+        };
+      </script>`;
+
+      const { code } = runPlugin(
+        input,
+        {},
+        {
+          processNestedWorklets: true,
+        }
```
Comment (tomekzaw) [inline __tests__/plugin.test.ts:1467]:
```suggestion
        { processNestedWorklets: true }
```

**Review by tomekzaw — APPROVED** [review]
👏

**Review by piaskowyk — APPROVED** [review]

**Comment by mrousavy** [comment]
Awesome work!! 👏

**Comment by Latropos** [comment]
@tjzel Do you want to add some documentation for this?

**Comment by tjzel** [comment]
@Latropos I never expected anyone to use this 😂

**Pull request merged**